feature/orchestrator/src/main/java/com/simprints/feature/orchestrator/usecases/steps/BuildStepsUseCase.kt

```diff
@@ -103,23 +82,126 @@ internal class BuildStepsUseCase @Inject constructor(
+            ageGroup,
+            subjectQuery = buildMatcherSubjectQuery(projectConfiguration, action),
+        )
+
+        is ActionRequest.VerifyActionRequest -> buildModalityCaptureAndMatchStepsForVerify(
+            action,
+            projectConfiguration,
+            ageGroup,
+        )
+
+        else -> emptyList()
+    }
+
+    private fun buildModalityCaptureAndMatchStepsForEnrol(
+        action: ActionRequest.EnrolActionRequest,
+        projectConfiguration: ProjectConfiguration,
+        ageGroup: AgeGroup? = null,
     ): List<Step> {
-        if (projectConfiguration.allowedAgeRanges().isEmpty()) {
-            return emptyList()
-        }
-        // Todo check if the action request contains the age parameter
+        val resolvedAgeGroup = ageGroup ?: ageGroupFromSubjectAge(action, projectConfiguration)
+
         return listOf(
-            Step(
-                id = StepId.SELECT_SUBJECT_AGE,
-                navigationActionId = R.id.action_orchestratorFragment_to_age_group_selection,
-                destinationId = SelectSubjectAgeGroupContract.DESTINATION,
-                payload = bundleOf()
+            buildModalityCaptureSteps(
+                projectConfiguration,
+                FlowType.ENROL,
+                resolvedAgeGroup,
+            ),
+            if (projectConfiguration.general.duplicateBiometricEnrolmentCheck) {
+                buildModalityMatcherSteps(
+                    projectConfiguration,
+                    FlowType.ENROL,
+                    resolvedAgeGroup,
+                    buildMatcherSubjectQuery(projectConfiguration, action),
+                    BiometricDataSource.fromString(action.biometricDataSource),
+                )
+            } else emptyList(),
+        ).flatten()
+    }
+
+    private fun buildModalityCaptureAndMatchStepsForIdentify(
+        action: ActionRequest.IdentifyActionRequest,
+        projectConfiguration: ProjectConfiguration,
+        ageGroup: AgeGroup? = null,
+        subjectQuery: SubjectQuery,
+    ): List<Step> {
+        val resolvedAgeGroup = ageGroup ?: ageGroupFromSubjectAge(action, projectConfiguration)
+
+        return listOf(
+            buildModalityCaptureSteps(
+                projectConfiguration,
+                FlowType.IDENTIFY,
+                resolvedAgeGroup,
+            ),
+            buildModalityMatcherSteps(
+                projectConfiguration,
+                FlowType.IDENTIFY,
+                resolvedAgeGroup,
+                subjectQuery,
+                BiometricDataSource.fromString(action.biometricDataSource),
             )
-        )
+        ).flatten()
+    }
 
+    private fun buildModalityCaptureAndMatchStepsForVerify(
+        action: ActionRequest.VerifyActionRequest,
+        projectConfiguration: ProjectConfiguration,
+        ageGroup: AgeGroup? = null,
+    ): List<Step> {
+        val resolvedAgeGroup = ageGroup ?: ageGroupFromSubjectAge(action, projectConfiguration)
```

Comment:
The problem is that all 3 functions with repeated code can be called from 2 places (which is why they were extracted in the first place). Was thinking of extracting this part in a helper function but IMO it would hardly improve things - less repetition but more lines of code.

Reply:
That said - if you can think of a good solution - it's more than welcome!